### ISSUE TITLE
coalesce / skip nulls mode for ReplacingMergeTree 

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#57240]
Quite often you need to enrich the data of the row at the later time. You may know all the dimension column values (i.e. all 'main columns' which are part of the table ORDER BY) but it can be problematic to find and (re)insert all the metrics which were inserted previously. 

In the trivial case

```
CREATE TABLE test ( id UInt64, metric1 Nullable(UInt64), metric2 Nullable(UInt64)) engine=ReplacingMergeTree ORDER BY id;

-- that insert knows nothing about metric2 value
INSERT INTO test (id, metric1) VALUES (1, 100); 

-- that insert knows nothing about metric1 value
INSERT INTO test (id, metric2) VALUES (1, 200);
```

Now we need to 'combine' / 'coalesce' those 2 rows - not fully replacing the old one with a new one, but collect non-null values of the metrics in one row (picking the most recent for every metric).

I.e. we expect to get 
```
SELECT * FROM test FINAL;

1,100,200
```

### Alternatives: 

The same/similar effect can be achieved by using `AggregatingFunction(any_last` or by `SimpleAggregateFunction`, but since that scenario is super-common, it would be better to add support for that into ReplacingMergeTree engine itself. 

Additional reasons. 

• Ergonomics: It should not require advanced SQL acrobatics in the schema or query design
• Schema evolution should be possible with reasonable effort. At least adding non-sorting key columns should be simple
• And of course query, finalization (if neded) and merge performance should be excellent.

### Proposal 

merge_tree setting called  `coalesce_previous_values` or `skip nulls`. When enabled logic of ReplacingMergeTree is changed to ignore nulls when do replacings.

**Comment by canhld94** [comment]
But from implementation PoV, ReplacingMergeAlgorithm cannot do what you described. What's your plan for implementation? Adding more logic to ReplacingMergeTree (I think this approach is quite complicated) or reuse AggregatingMergeAlgorithm when the table has coalesce settings?

**Comment by UnamedRus** [comment]
IMO, we should add mode (via mergetree setting: `default_aggregating_function`) to AggregatingMergeTree, which will use `any` aggregate functions for columns, which are not in ORDER BY key, but don't have `*AggregateFunction*` data type. 
Current behavior is quite similar, but it doesn't ignore `null` during merge, instead it keep them.

**Comment by aadant** [comment]
`anyLast` also makes sense

**Comment by filimonov** [comment]
I guess guys also wanted to use delete functionality in Replacing?

**Comment by zvonand** [comment]
Maybe introduce a standalone table engine for this purpose (in case this is a really common scenario)? Something like CoalescingReplacingMergeTree.  
IMO the proposed behavior doesn't really align well with what ReplacingMergeTree is doing -- it is not really _replacing_.

**Comment by BorisTyshkevich** [comment]
The original idea fits better to AggregatingMT logic but is easy to use from the user's point of view (no need to specify AggregatingFunctions for every column).  We already have SummingMT as the descendant of AggregatingMT with fixed AggregatingFunction - sum() for any numeric non-ORDER BY column.  

We could make the same trick for CoalescingMergeTree - make it a descendant of AggregatingMT with a fixed AggregatingFunction - argMax(..., version).

**Comment by zvonand** [comment]
Yes, my idea came from understanding that we already have SummingMergeTree that is quite similar. And the initial proposal does not really fit ReplacingMergeTree.

My only concern is that it would be an overkill -- first there was an engine for `sum`, now we would add another one for `anyLast`, then someone will need another one for `max`.

However, if this is a **_really_** common usage scenario, it may be a good idea to make it a standalone engine. I don't know how useful it would be from users' point of view, and if this approach will be accepted and merged.

**Comment by UnamedRus** [comment]
> AggregatingFunction - argMax(..., version).

Just regular AggregatingFunction argMax have big overhead of storing version per column, so it should be some more smart implementation, and regular argMax function can't be reused. 


TBH, 

I'd prefer to reduce amount of table types, especially as they only differ in merge/SELECT FINAL behavior only. 
And everything else almost the same. and make this merge behavior selector into merge tree setting

**Comment by oplehto** [comment]
> Just regular AggregatingFunction argMax have big overhead of storing version per column, so it should be some more smart implementation, and regular argMax function can't be reused.

If implemented smartly, one advantage is also that SELECT FINAL could be quite fast: Any rows that do not have any NULLs could be considered guaranteed to be finalized and skip a lot of the merge logic.

**Comment by zvonand** [comment]
Currently, I see the following options:

* **Implement this as a MergeTree setting to be used with ReplacingMergeTree**
  - It does not quite fit ReplacingMergeTree logic and purpose -- it is actually not **_replacing_** in case this setting is enabled
  - RMT was not initially designed for this, so this may require a lot of changes
  - But it will be an easy-to-use solution
  
* **Implement this as another MergeTree family engine** (e.g. AnyMergeTree)
  - Rather easy to implement (SummingMergeTree can be taken as a good example) and easy to use
  - This may be considered an overkill. However, if this is a very common scenario, it may be a good solution
  - Otherwise I don't think it is a good idea to create a new table engine for every particular use case. 
  
* https://github.com/ClickHouse/ClickHouse/issues/57240#issuecomment-1827751621
  - May be not as easy to use as previous two options
  
Which of this do you think is more viable?

**Comment by karthikvt26** [comment]
I just posted a comment on the ClickHouse [slack channel](https://clickhousedb.slack.com/archives/CU478UEQZ/p1715172399313339) about a possible use case of this if not via any other way to achieve the same in a better way. Thanks

----- 
OpenTelemetry span captures business specific context information about
- Operation being performed
- Identify user's resources
- Other business-specific context captured
In each span as part of the distributed tracing story goes

In microservices, each service in the request path could emit critical business data that you would like to collect and aggregate based on say TraceID. This ticket on the community repo closely resembles the problem I am trying to describe. You could achieve something similar via this but this approach IIUC won't help filter the info based on the aggregated column faster when you are looking at billions of records.

**Comment by jasine** [comment]
> Currently, I see the following options:
> 
> * **Implement this as a MergeTree setting to be used with ReplacingMergeTree**
>   
>   * It does not quite fit ReplacingMergeTree logic and purpose -- it is actually not **_replacing_** in case this setting is enabled
>   * RMT was not initially designed for this, so this may require a lot of changes
>   * But it will be an easy-to-use solution
> * **Implement this as another MergeTree family engine** (e.g. AnyMergeTree)
>   
>   * Rather easy to implement (SummingMergeTree can be taken as a good example) and easy to use
>   * This may be considered an overkill. However, if this is a very common scenario, it may be a good solution
>   * Otherwise I don't think it is a good idea to create a new table engine for every particular use case.
>   * **(Currently in progress)** Maybe implement this as some "lightweight" AggregatingMT that can be parametrized by one stateless aggregate function (applicable to all aggregated columns) that would be more user-friendly than plain AggregatingMT. See [DRAFT: StatelessAggregatingMergeTree #59777](https://github.com/ClickHouse/ClickHouse/pull/59777) for details.
> * **Use AggregatingMergeTree**
>   > IMO, we should add mode (via mergetree setting: default_aggregating_function) to AggregatingMergeTree, which will use any aggregate functions for columns, which are not in ORDER BY key, but don't have _AggregateFunction_ data type.
>   > Current behavior is quite similar, but it doesn't ignore null during merge, instead it keep them.
>   
>   
>   See [Add `default_aggregate_function` parameter for AggregatingMergeTree #62707](https://github.com/ClickHouse/ClickHouse/pull/62707)

In our scenario, the first solution is more suitable

**Comment by zvonand** [comment]
> In our scenario, the first solution is more suitable

Understandable, but the third (#62707) option is still more "natural".  
Enabling this for one particular use case would be strange, it is better to allow user to choose the aggregate function that he needs.
